lib/src/test/java/io/ably/lib/http/HttpTest.java

```diff
@@ -0,0 +1,447 @@
+
+        /* Verify there was a fallback with first call */
+        assertThat(argument.getValue().getHost().matches(PATTERN_HOST_FALLBACK), is(true));
+
+        /* Update behavior to perform a call without a fallback */
+        argument = ArgumentCaptor.forClass(URL.class);
+        doReturn(null)
+                .when(http).httpExecute(argument.capture(), anyString(), aryEq(new Param[0]),
+                any(Http.RequestBody.class), anyBoolean(), any(Http.ResponseHandler.class));
+
+        http.ablyHttpExecute("", "", new Param[0], new Param[0],
+                Mockito.mock(Http.RequestBody.class), Mockito.mock(Http.ResponseHandler.class));
+
+        /* Verify second call was called with http host */
+        assertThat(argument.getValue().getHost().equals(hostExpected), is(true));
+    }
+
+    /**
+     * <p>
+     * Validates {@code Http} is throwing an exception,
+     * when connection to host failed more than allowed count ({@code Defaults.HTTP_MAX_RETRY_COUNT})
+     * </p>
+     * <p>
+     * Spec: -
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_excessivefallback () throws AblyException {
+        ClientOptions options = new ClientOptions();
+        Http http = Mockito.spy(new Http(options, null));
+
+        ArgumentCaptor<URL> argument = ArgumentCaptor.forClass(URL.class);
+        int excessiveFallbackCount = options.httpMaxRetryCount + 1;
+
+        /* Partially mock http */
+        doAnswer(new GrumpyAnswer(excessiveFallbackCount, new Http.HostFailedException(new Exception()), ""))
+                .when(http).httpExecute(argument.capture(), anyString(), aryEq(new Param[0]),
+                any(Http.RequestBody.class), anyBoolean(), any(Http.ResponseHandler.class));
+
+
+        /* Verify
+         *   - ably exception with 404 status code is thrown
+         */
+        ErrorInfo expectedErrorInfo = new ErrorInfo("Connection failed; no host available", 404, 80000);
+        thrown.expect(AblyException.class);
+        thrown.expect(new ErrorInfoMatcher(expectedErrorInfo));
+
+        http.ablyHttpExecute("", "", new Param[0], new Param[0],
+                Mockito.mock(Http.RequestBody.class), Mockito.mock(Http.ResponseHandler.class));
+    }
+
+    /**
+     * <p>
+     * Validates {@code Http} is performing a fallback,
+     * when api returns a response code between 500 and 504
+     * </p>
+     * <p>
+     * Spec: RSC15d
+     * </p>
+     *
+     * @throws Exception
+     */
+    @Test
+    public void http_execute_throws_500() throws MalformedURLException, AblyException {
+        Http http = new Http(new ClientOptions(), null);
+        Http.RequestBody requestBody = new Http.ByteArrayRequestBody(new byte[0], NanoHTTPD.MIME_PLAINTEXT);
+        URL url = new URL("http://localhost:" + server.getListeningPort() + "/status/500");
+
+        /* Verify
+         *   - HostFailedException is thrown */
+        thrown.expect(Http.HostFailedException.class);
```

Comment:
I know that, but `is performing a fallback` is the part I was talking about, it's not performing any fallback in this test